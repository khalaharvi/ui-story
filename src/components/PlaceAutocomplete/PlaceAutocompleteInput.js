import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';
import _get from 'lodash.get';

import { Plus as PlusIcon } from 'icons';
import { cleanButton } from '../../styles';

const InputContainer = styled('div')({
  position: 'relative'
});

const InputButton = styled('button')(({ theme, hidden }) => ({
  ...cleanButton,
  display: hidden ? 'none' : 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  position: 'absolute',
  right: 5,
  top: '50%',
  transform: 'translateY(-50%)',
  height: 40,
  width: 40,

  ':hover svg': {
    fill: _get(theme, 'text.primary')
  },

  svg: {
    display: 'block',
    width: 14,
    height: 14,
    fill: _get(theme, 'text.secondary')
  }
}));

const PlaceAutocompleteInput = ({
  canReset,
  onReset,
  ...props
}) => (
  <InputContainer>
    <input {...props} />
    <InputButton type="button" hidden={!canReset} onClick={onReset}>
      <PlusIcon rotate />
    </InputButton>
  </InputContainer>
);

PlaceAutocompleteInput.propTypes = {
  onReset: PropTypes.func,
  canReset: PropTypes.bool
};

PlaceAutocompleteInput.defaultProps = {
  canReset: false
};

export default PlaceAutocompleteInput;
