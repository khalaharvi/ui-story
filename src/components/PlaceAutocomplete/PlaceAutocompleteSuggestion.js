import React from 'react';
import styled from 'react-emotion';
import PropTypes from 'prop-types';

const PlaceName = styled('span')({
  overflow: 'hidden',
  display: 'block',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap'
});

const PlaceAutocompleteSuggestion = ({ place }) => (
  <div><PlaceName>{place.name}</PlaceName></div>
);

PlaceAutocompleteSuggestion.propTypes = {
  place: PropTypes.shape({
    name: PropTypes.string.isRequired
  })
};

export default PlaceAutocompleteSuggestion;
