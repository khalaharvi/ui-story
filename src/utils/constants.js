export const PRIMARY_CATEGORY_KEY = 'categories.lvl0';
export const SECONDARY_CATEGORY_KEY = 'categories.lvl1';
export const TERTIARY_CATEGORY_KEY = 'categories.lvl2';
