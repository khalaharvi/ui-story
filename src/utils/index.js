export { default as createClassNames } from './createClassNames';
export { default as getDisplayName } from './getDisplayName';
export { default as capitalizeFirstLetter } from './capitalizeFirstLetter';
export { default as slugify } from './slugify';
