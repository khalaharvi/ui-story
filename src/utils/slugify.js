export default (text) => {
  // Use hash map for special characters
  const specialChars = {"à":'a',"ä":'a',"á":'a',"â":'a',"æ":'a',"å":'a',"ë":'e',"è":'e',"é":'e', "ê":'e',"î":'i',"ï":'i',"ì":'i',"í":'i',"ò":'o',"ó":'o',"ö":'o',"ô":'o',"ø":'o',"ù":'o',"ú":'u',"ü":'u',"û":'u',"ñ":'n',"ç":'c',"ß":'s',"ÿ":'y',"œ":'o',"ŕ":'r',"ś":'s',"ń":'n',"ṕ":'p',"ẃ":'w',"ǵ":'g',"ǹ":'n',"ḿ":'m',"ǘ":'u',"ẍ":'x',"ź":'z',"ḧ":'h',"·":'-',"/":'-',"_":'-',",":'-',":":'-',";":'-'}; // eslint-disable-line

  return text.toString().toLowerCase()
    .replace(/\s+/g, '-') // Replace spaces with -
    .replace(/./g, target => specialChars[target] || target) // Replace special characters using the hash map
    .replace(/&/g, '-and-') // Replace & with 'and'
    .replace(/[^\w-]+/g, '') // Remove all non-word chars
    .replace(/--+/g, '-') // Replace multiple - with single -
    .replace(/^-+/, '') // Trim - from start of text
    .replace(/-+$/, ''); // Trim - from end of text
};
