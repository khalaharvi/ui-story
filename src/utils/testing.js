import { object } from 'prop-types';
import { channel, createBroadcast } from 'emotion-theming';
import { shallow, mount } from 'enzyme';
import theme from '@dering-hall/dh-theme';

const getOptions = (options = {}) => {
  const broadcast = createBroadcast(theme);

  return {
    context: {
      [channel]: broadcast
    },
    childContextTypes: {
      [channel]: object
    },
    ...options
  };
};

export const mountWithTheme = (
  Component,
  options = {}
) => mount(Component, { ...getOptions(options) });

export const shallowWithTheme = (
  Component,
  options = {}
) => shallow(Component, { ...getOptions(options) });

export default {
  mountWithTheme,
  shallowWithTheme
};
