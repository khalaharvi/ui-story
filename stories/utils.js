/* eslint-disable import/prefer-default-export */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { InstantSearch } from 'react-instantsearch/dom';
import { connectStateResults } from 'react-instantsearch/connectors';
import { withMedia, MediaQueryProvider } from 'react-media-query-hoc';
import { withTests as withTestsAddon } from '@storybook/addon-jest';
import { ThemeProvider } from 'emotion-theming';
import dhTheme from '@dering-hall/dh-theme';
import testsResults from '../.jest-test-results.json';

const Wrap = ({ index, children }) => (
  <InstantSearch
    appId={process.env.ALGOLIA_ID}
    apiKey={process.env.ALGOLIA_API_KEY}
    indexName={index}
  >
    {children}
  </InstantSearch>
);

Wrap.propTypes = {
  index: PropTypes.string,
  children: PropTypes.node
};

Wrap.defaultProps = {
  index: process.env.ALGOLIA_PRODUCT_INDEX
};

const WrapThemeProvider = ({ children, theme }) => (
  <ThemeProvider theme={theme}>
    <Fragment>
      {children}
    </Fragment>
  </ThemeProvider>
);

WrapThemeProvider.propTypes = {
  theme: PropTypes.object // eslint-disable-line
};

WrapThemeProvider.defaultProps = {
  theme: dhTheme
};

const SearchStateDebug = connectStateResults(({ searchState }) => (
  <div style={{ margin: '1rem 0', fontFamily: 'monospace' }}>
    <pre
      style={{
        background: '#f6f8fa',
        fontSize: '.65rem',
        padding: '.5rem'
      }}
    >
      <strong>searchState</strong> = {JSON.stringify(searchState, null, '\t')}
    </pre>
  </div>
));

const customQueries = {
  small: 'screen and (max-width: 730px)',
  medium: 'screen and (max-width: 970px)',
  large: 'screen and (max-width: 1240px)',
  xlarge: 'screen and (min-width: 1240px)'
};

const MediaQueriesWrapper = ({ children }) => (
  <MediaQueryProvider queries={customQueries}>
    {children}
  </MediaQueryProvider>
);

const OnSmallOnly = withMedia(({ media, children }) => {
  if (!media.small) {
    return null;
  }

  return (
    <Fragment>
      {children}
    </Fragment>
  );
});

const withTests = withTestsAddon({
  results: testsResults
});

export {
  SearchStateDebug,
  MediaQueriesWrapper,
  OnSmallOnly,
  Wrap,
  WrapThemeProvider,
  withTests
};
/* eslint-enable */
