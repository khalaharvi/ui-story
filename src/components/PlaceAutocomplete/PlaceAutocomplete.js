import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import _debounce from 'lodash.debounce';
import Autosuggest from 'react-autosuggest';
import { css } from 'react-emotion';
import { withTheme } from 'emotion-theming';
import _get from 'lodash.get';

import { createClassNames } from 'utils';
import PlaceAutocompleteSuggestion from './PlaceAutocompleteSuggestion';
import PlaceAutocompleteInput from './PlaceAutocompleteInput';
import {
  INPUT_X_PADDING,
  INPUT_LINE_HEIGHT,
  inputBgColor,
  inputBorderColor,
  inputPlaceholderColor
} from '../Form/FormInput';

export const cx = createClassNames('place-autocomplete');

const getReactAutosuggestClassnames = (theme, error) => ({
  container: css({
    position: 'relative'
  }),
  input: css({
    width: '100%',
    padding: `0 70px 0 ${INPUT_X_PADDING}px`,
    backgroundColor: inputBgColor(theme),
    border: `1px solid ${inputBorderColor(theme, error)}`,
    lineHeight: `${INPUT_LINE_HEIGHT}px`,
    outline: 0,

    '::placeholder': {
      color: inputPlaceholderColor(theme)
    }
  }),
  suggestionsContainer: css({
    position: 'absolute',
    top: '100%',
    left: 0,
    zIndex: 1,
    width: '100%',

    backgroundColor: inputBgColor(theme),
    borderRight: `1px solid ${inputBorderColor(theme)}`,
    borderLeft: `1px solid ${inputBorderColor(theme)}`
  }),
  suggestionsList: css({
    margin: 0,
    padding: 0,
    listStyleType: 'none'
  }),
  suggestion: css({
    padding: `0 ${INPUT_X_PADDING}px`,
    lineHeight: `${INPUT_LINE_HEIGHT + 1}px`,
    borderBottom: `1px solid ${inputBorderColor(theme)}`,
    cursor: 'pointer'
  }),
  suggestionHighlighted: css({
    backgroundColor: _get(theme, 'colors.secondary.beige'),
    color: _get(theme, 'text.primary')
  })
});

class PlaceAutocomplete extends Component {
  constructor(props) {
    super(props);

    this.state = {
      places: [],
      inputValue: props.defaultValue || ''
    };
  }

  componentDidMount() {
    if (!window.google) {
      throw new Error('Google Maps JavaScript API library must be loaded.');
    }

    if (!window.google.maps.places) {
      throw new Error('Google Maps Places library must be loaded.');
    }

    this.autocompleteService = new window.google.maps.places.AutocompleteService();
    this.autocompleteOK = window.google.maps.places.PlacesServiceStatus.OK;
    this.autocompleteNoResults = window.google.maps.places.PlacesServiceStatus.ZERO_RESULTS;

    this.geocoderService = new window.google.maps.Geocoder();
    this.geocoderOK = window.google.maps.GeocoderStatus.OK;
  }

  onSuggestionsClearRequested = () => {
    this.setState({
      places: []
    });
  };

  onSuggestionsFetchRequested = ({ value }) => {
    this.fetchPlacesPredictions(value);
  };

  onSuggestionSelected = (event, {
    suggestion,
    suggestionValue,
    suggestionIndex,
    method
  }) => {
    if (method === 'enter') {
      event.preventDefault();
    }

    // Don't allow selection of 'No Results.'
    if (suggestion.placeId === this.autocompleteNoResults) {
      this.setState({ inputValue: '' });

      return;
    }

    const { onPlaceSelected, withoutGeoCoords } = this.props;

    if (onPlaceSelected) {
      if (withoutGeoCoords) {
        onPlaceSelected(event, {
          place: suggestion,
          placeValue: suggestionValue,
          placeIndex: suggestionIndex,
          method
        });
      } else {
        this.getLatLng(suggestion.placeId).then((coordinates) => {
          const place = Object.assign({}, suggestion, {
            _geoloc: coordinates
          });

          onPlaceSelected(event, {
            place,
            placeValue: suggestionValue,
            placeIndex: suggestionIndex,
            method
          });
        });
      }
    }
  };

  onInputChange = (e, { newValue }) => {
    const { onChange } = this.props;

    this.setState({ inputValue: newValue });

    if (onChange) {
      onChange(e);
    }
  };

  onInputReset = (e) => {
    const { onReset } = this.props;

    this.setState({ inputValue: '' });

    if (onReset) {
      onReset(e);
    }
  };

  // Adapted from https://github.com/kenny-hibino/react-places-autocomplete/blob/v6.1.3/src/utils.js
  getLatLng = placeId => new Promise((resolve, reject) => {
    this.geocoderService.geocode({ placeId }, (results, status) => {
      if (status !== this.geocoderOK) {
        reject(status);
      }

      try {
        const latLng = {
          lat: results[0].geometry.location.lat(),
          lng: results[0].geometry.location.lng()
        };

        resolve(latLng);
      } catch (e) {
        reject(e);
      }
    });
  });

  fetchPlacesPredictions = (value) => {
    if (value.length > 0) {
      this.autocompleteService.getPlacePredictions(
        {
          types: ['(cities)'],
          input: value
        },
        this.placesAutocompleteCallback
      );
    }
  };

  placesAutocompleteCallback = (predictions, status) => {
    if (status !== this.autocompleteOK && status !== this.autocompleteNoResults) {
      this.props.onError(status, this.onSuggestionsClearRequested);

      return;
    }

    if (status === this.autocompleteNoResults) {
      this.setState({
        places: [{
          name: 'No results.',
          placeId: this.autocompleteNoResults
        }]
      });

      return;
    }

    this.setState({
      places: predictions.map(p => ({
        name: p.description,
        placeId: p.place_id
      }))
    });
  };

  render() {
    const {
      className,
      placeholder,
      theme,
      id,
      onBlur,
      error
    } = this.props;

    const { inputValue, places } = this.state;

    const reactAutosuggestClassnames = getReactAutosuggestClassnames(theme, error);

    return (
      <div className={classnames(cx(''), className)}>
        <Autosuggest
          ref={(autosuggest) => {
            if (autosuggest !== null) {
              this.input = autosuggest.input;
            }
          }}
          highlightFirstSuggestion
          inputProps={{
            onReset: this.onInputReset,
            canReset: (inputValue !== ''),
            onChange: this.onInputChange,
            onBlur,
            placeholder,
            value: inputValue,
            id
          }}
          getSuggestionValue={place => place.name}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested}
          onSuggestionsFetchRequested={_debounce(this.onSuggestionsFetchRequested, 200)}
          onSuggestionSelected={this.onSuggestionSelected}
          renderInputComponent={PlaceAutocompleteInput}
          renderSuggestion={place => <PlaceAutocompleteSuggestion place={place} />}
          suggestions={places}

          theme={reactAutosuggestClassnames}
          focusInputOnSuggestionClick={false}
        />
      </div>
    );
  }
}

PlaceAutocomplete.propTypes = {
  className: PropTypes.string,
  onError: PropTypes.func,
  onChange: PropTypes.func,
  onPlaceSelected: PropTypes.func,
  onReset: PropTypes.func,
  placeholder: PropTypes.string,
  defaultValue: PropTypes.string,
  withoutGeoCoords: PropTypes.bool,
  id: PropTypes.string
};

PlaceAutocomplete.defaultProps = {
  placeholder: 'New York, NY',
  withoutGeoCoords: false,
  onChange: () => {},
  onError: (status, _clearSuggestions) => {
    console.error(
      '[<PlaceAutocomplete />] An error happened when fetching data from Google Maps API.\nPlease check the docs here (https://developers.google.com/maps/documentation/javascript/places#place_details_responses)\nStatus: ',
      status
    );

    _clearSuggestions();
  }
};

export default withTheme(PlaceAutocomplete);
