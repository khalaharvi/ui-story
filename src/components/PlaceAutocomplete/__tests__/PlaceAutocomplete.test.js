import React from 'react';
import { shallowWithTheme } from 'utils/testing';

import PlaceAutocomplete from '../PlaceAutocomplete';

describe('<PlaceAutocomplete />', () => {
  it('renders correctly', () => {
    const wrapper = shallowWithTheme(<PlaceAutocomplete />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with custom placeholder', () => {
    const wrapper = shallowWithTheme(<PlaceAutocomplete placeholder="From..." />);

    expect(wrapper).toMatchSnapshot();
  });
});
