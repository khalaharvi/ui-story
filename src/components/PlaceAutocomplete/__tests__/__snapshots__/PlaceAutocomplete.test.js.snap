// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaceAutocomplete /> renders correctly 1`] = `
<PlaceAutocomplete
  onChange={[Function]}
  onError={[Function]}
  placeholder="New York, NY"
  withoutGeoCoords={false}
/>
`;

exports[`<PlaceAutocomplete /> renders correctly with custom placeholder 1`] = `
<PlaceAutocomplete
  onChange={[Function]}
  onError={[Function]}
  placeholder="From..."
  withoutGeoCoords={false}
/>
`;
